syncing props and state is usually a bad idea

jyash97 opened the pull request:
Update over the issue #337

bvaughn commented on content/docs/faq-state.md:
```diff
@@ -18,6 +18,14 @@ Here are some good resources for further reading on when to use `props` vs `stat
+
+In rare cases, it’s okay to initialize state based on [`props`](/docs/components-and-props.html). This effectively “forks” the props and sets the state with the initial props.
+
+Beware of this pattern, as state won't be up-to-date with any props update. Instead of syncing props to state, you often want to [lift the state up](/docs/lifting-state-up.html) instead.
```
Link to anchor tag:
```
[lift the state up](/docs/lifting-state-up.html#lifting-state-up)
```

bvaughn commented on content/docs/faq-state.md:
```diff
@@ -18,6 +18,14 @@ Here are some good resources for further reading on when to use `props` vs `stat
+
+Beware of this pattern, as state won't be up-to-date with any props update. Instead of syncing props to state, you often want to [lift the state up](/docs/lifting-state-up.html) instead.
+
+If you "fork" props by using them for state, you might also want to implement [`componentWillReceiveProps(nextProps)`](#componentwillreceiveprops) to keep the state up-to-date with them. But lifting state up is often easier and less bug-prone.
```
This link is wrong. Should point to:
```
[`componentWillReceiveProps(nextProps)`](/docs/react-component.html#componentwillreceiveprops)
```

jyash97 commented:
Hey @bvaughn have updated the links and added a example

jyash97 commented:
#337

gaearon commented on content/docs/faq-state.md:
```diff
@@ -18,6 +18,14 @@ Here are some good resources for further reading on when to use `props` vs `stat
 * [ReactJS: Props vs. State](http://lucybain.com/blog/2016/react-state-vs-pros/)
 
+### Is synching state and props a bad idea?
+
```
I think it's not very clear from this description what "syncing" means.

I think we should show a three-line code example first and *then* explain why it's problematic.

jyash97 commented:
@gaearon  
I have added the code  example

gaearon commented:
This shows the “good” example. I think we should show the “bad” example and how to fix it first.

jyash97 commented:
I do not get the context of "good" and "bad" example but for fixing it we can add componentWillReceiveProps.
@gaearon

gaearon commented:
Bad code is when somebody does

```
constructor(props) {
  super(props);
  this.state = { isOn: props.isOn };
}

handleClick() {
  this.setState(prevState => ({
    isOn: !prevState.isOn
  });
}

render() {
  return (
    <button onClick={() => this.handleClick()>
      The switch is {this.state.isOn ? 'on' : 'off'}
    </button>
  );
}
``` 

“Good” example removes the state and shows how to lift it up to the parent instead.

jyash97 commented:
Okay Got it, will take the same example and add description,

jyash97 commented:
Hey @gaearon @bvaughn 
I have added the _bad example_ with its solution and I have added the full code. Please reduce its length as I am not getting which code would be right to display.
Hoping this time I did what was required.

gaearon commented on content/docs/faq-state.md:
```diff
@@ -18,6 +18,75 @@ Here are some good resources for further reading on when to use `props` vs `stat
+
+In rare cases, it's okay to initialize state based on props. One valid use to store props in state is to be able to know its previous values, because props can change over time. This effectively "forks" the props and sets the state with the initial props.
+
+Here's an example of what not to do:
```
It's a bit confusing we start with "in rare cases it's okay" but then the example shows when it's not okay. Let's change to start with the "bad" section and only mention the good use case at the end.

jyash97 commented:
@gaearon I  have deleted some lines which might confuse users.

jyash97 commented:
@gaearon @bvaughn  I have removed the lines which might confuse users.

jyash97 commented:
@gaearon  Should we close #337 or any update needed?

gaearon commented:
Why did you close? We don't currently have capacity to review all PRs but we'd get to it eventually.

jyash97 commented:
Let me reopen it again

rickhanlonii commented:
Sorry for the long delay here. Since this PR was created against the old site I'm going to close this, but I think this use case is covered in the new docs for hooks in You Might Not Need an Effect: https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes
